src/rm_topk.c

```diff
@@ -0,0 +1,245 @@
+
+    size_t itemlen;
+    long long res;
+    RedisModule_ReplyWithArray(ctx, argc - 2);
+    for(int i = 2; i < argc; ++i) {
+        const char *item = RedisModule_StringPtrLen(argv[i], &itemlen);
+        res = TopK_Query(topk, item, itemlen);
+        RedisModule_ReplyWithLongLong(ctx, res);
+    }
+    
+    return REDISMODULE_OK;
+}
+
+int TopK_Count_Cmd(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
+   RedisModule_AutoMemory(ctx);    
+    if (argc < 3)
+        return RedisModule_WrongArity(ctx);
+    
+    TopK *topk;
+    if(GetTopKKey(ctx, argv[1], &topk, REDISMODULE_READ) != REDISMODULE_OK)
+        return REDISMODULE_ERR;
+
+    size_t itemlen;
+    long long res;
+    RedisModule_ReplyWithArray(ctx, argc - 2);
+    for(int i = 2; i < argc; ++i) {
+        const char *item = RedisModule_StringPtrLen(argv[i], &itemlen);
+        res = TopK_Count(topk, item, itemlen);
+        RedisModule_ReplyWithLongLong(ctx, res);
+    }
+    
+    return REDISMODULE_OK;
+}
+
+int TopK_List_Cmd(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
+    RedisModule_AutoMemory(ctx);
+    if (argc != 2)
+        return RedisModule_WrongArity(ctx);
+
+    TopK *topk = NULL;
+    if (GetTopKKey(ctx, argv[1], &topk, REDISMODULE_READ) != REDISMODULE_OK) {
+        return REDISMODULE_OK;
+    }
+
+    char **heapList = TOPK_CALLOC(topk->k, (sizeof(char *)));
+    uint32_t qty = TopK_List(topk, heapList);
+    RedisModule_ReplyWithArray(ctx, qty);
+    for(int i = 0; i < qty; ++i) {
+        if(heapList[i]) {
+            RedisModule_ReplyWithSimpleString(ctx, heapList[i]);
+        }
+    }
+    TOPK_FREE(heapList);
+
+    return REDISMODULE_OK;
+}
+
+int TopK_Info_Cmd(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
+    RedisModule_AutoMemory(ctx);
+    if (argc != 2)
+        return RedisModule_WrongArity(ctx);
+
+    TopK *topk = NULL;
+    if (GetTopKKey(ctx, argv[1], &topk, REDISMODULE_READ) != REDISMODULE_OK) {
+        return REDISMODULE_OK;
+    }
+
+    RedisModule_ReplyWithArray(ctx, 4 * 2);
+    RedisModule_ReplyWithSimpleString(ctx, "k");
+    RedisModule_ReplyWithLongLong(ctx, topk->k);
+    RedisModule_ReplyWithSimpleString(ctx, "width");
+    RedisModule_ReplyWithLongLong(ctx, topk->width);
+    RedisModule_ReplyWithSimpleString(ctx, "depth");
+    RedisModule_ReplyWithLongLong(ctx, topk->depth);
+    RedisModule_ReplyWithSimpleString(ctx, "decay");
+    RedisModule_ReplyWithDouble(ctx, topk->decay);
+
+    return REDISMODULE_OK;
+}
+
+/**************** Module functions *********************************/
+
+void TopKRdbSave(RedisModuleIO *io, void *obj) {
+    TopK *topk = obj;
+    RedisModule_SaveUnsigned(io, topk->k);
+    RedisModule_SaveUnsigned(io, topk->width);
+    RedisModule_SaveUnsigned(io, topk->depth);
+    RedisModule_SaveDouble(io, topk->decay);
+    RedisModule_SaveStringBuffer(io, (const char *)topk->data,
+                                 topk->width * topk->depth * sizeof(Bucket));
```

Comment:
There seems to be a potential memory error here. To illustrate in the simplest case:
`$ valgrind redis-server --loadmodule ./rebloom.so`
`$ redis-cli topk.reserve topk 20 50 5 0.9`
`$ redis-cli save`

Logs the output:
```
==15047== Conditional jump or move depends on uninitialised value(s)
==15047==    at 0x14790C: lzf_compress (lzf_c.c:153)
==15047==    by 0x15F505: rdbSaveLzfStringObject.part.2 (rdb.c:349)
==15047==    by 0x15F90C: rdbSaveLzfStringObject (rdb.c:346)
==15047==    by 0x15F90C: rdbSaveRawString (rdb.c:421)
==15047==    by 0x1B1873: RM_SaveStringBuffer (module.c:3153)
==15047==    by 0x83E78E3: TopKRdbSave (rm_topk.c:193)
==15047==    by 0x160AAD: rdbSaveObject (rdb.c:978)
==15047==    by 0x1614FD: rdbSaveKeyValuePair (rdb.c:1041)
==15047==    by 0x161BBC: rdbSaveRio (rdb.c:1145)
==15047==    by 0x16210A: rdbSave (rdb.c:1244)
==15047==    by 0x1650CE: saveCommand (rdb.c:2396)
==15047==    by 0x13FC36: call (server.c:2437)
==15047==    by 0x14052D: processCommand (server.c:2729)
```

Running the `topk.py` tests under Valgrind report further errors running count queries after a DEBUG RELOAD, but the root cause is probably here.

Reply:
This is a 'feature' of the module itself.
Opened an issue, but shouldn't effect this PR.

Reply:
:+1: Noted.

This issue should be investigated separately soon, though.